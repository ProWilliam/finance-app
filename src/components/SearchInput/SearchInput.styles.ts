import { StyleSheet } from 'react-native';
import theme from '../../theme/theme';

const StyleSearchInput = StyleSheet.create({

  containerSearch : {
    width: `${theme.withScreen.medium}%`,
    paddingVertical: 10,
    margin: 40,
    backgroundColor: theme.light.backgroundPrimary,
  },
  input: {
    width: `${theme.withScreen.big}%`,
    height: 60,
    fontSize: 16,
    paddingHorizontal: 20,
    borderWidth: 1,
    borderColor: theme.light.border,
    borderRadius: 8,
    overflow: 'hidden',
  }
});

export default StyleSearchInput;