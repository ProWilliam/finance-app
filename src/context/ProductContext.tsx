import React, { createContext, useState, useContext } from 'react';

// Types
import { GlobalProviderProps, ProductContextInterface, Product } from './ProductContext.type';


// Default State Context by products.
const defaultStateProduct = {
  products: [{
    id: '',
    name: ''
  }],
  setProducts: (product: Product[]) => {}

} as ProductContextInterface

const MyContextProducts = createContext<ProductContextInterface>(defaultStateProduct);

export const useMyContextProduct = () => {
  const context = useContext(MyContextProducts);

  if(!context) {
    throw new Error('useMycontext must be used within a MyContextProvider');
  }

  return context;
}

export const ProductProvider: React.FC<GlobalProviderProps> = ({ children }) => {
  const [ products, setProducts ] = useState<Product[]>([]);

  return (
    <MyContextProducts.Provider value={{ products, setProducts }}>
      {children}
    </MyContextProducts.Provider>
  );
};