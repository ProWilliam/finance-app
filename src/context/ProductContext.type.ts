import { ReactNode, Dispatch, SetStateAction } from 'react';

export type Product = {
  id: string,
  name: string
}

export interface ProductContextInterface {
  products: Product[];
  setProducts: Dispatch<SetStateAction<Product[]>>;
}

export interface GlobalProviderProps {
  children: ReactNode;
}