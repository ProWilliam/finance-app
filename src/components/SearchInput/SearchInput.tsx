import React, { useState } from 'react';
import { TextInput, View } from 'react-native';

// Hook
import useApi from '../../hooks/useApi';

// Component
import SearchInputProps from '../../types/components/SearchInput.types';

// Styles and Config
import style from './SearchInput.styles';
import config from '../../../app.config';
import { useMyContextProduct } from '../../context/ProductContext';

const SearchInput: React.FC<SearchInputProps> = ({ placeholder}) => {

  const [searchText, setSearchText] = useState<string>('');
  const { setProducts } = useMyContextProduct();
  
  const url = `${config.extra.productUrl + searchText}`
  const { data, refetch } = useApi(url);

  const handleSearch = async () => {

    await refetch();

    if (searchText === ''){
      setProducts(data.data)
    } else {
      setProducts([{
        id: data.id,
        name: data.name
      }])
    }
  }

  return (
    <View style={style.containerSearch}>
      <TextInput
        keyboardType='number-pad' 
        style={style.input}
        placeholder={placeholder}
        placeholderTextColor='black'
        value={searchText}
        onChangeText={(text) => setSearchText(text)}
        onSubmitEditing={handleSearch}
        returnKeyType='done'
      />
    </View>
  )
}

export default SearchInput