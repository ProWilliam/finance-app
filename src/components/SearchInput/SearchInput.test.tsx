import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { TextInput } from 'react-native';

import { useMyContextProduct } from '../../context/ProductContext';

import SearchInput from './SearchInput';

const MockuseApi = jest.fn();

jest.mock('../../hooks/useApi', () => ({
  __esModule: true,
  default: () => ({
    appNavigation: MockuseApi,
  }),
}));

jest.mock('../../context/ProductContext', () => ({
  useMyContextProduct: jest.fn(),
}));

describe('<SearchInput />', () => {

  const mockRefetch = jest.fn();
  const mockSetProducts = jest.fn();

  beforeEach(() => {
    MockuseApi.mockReturnValue({
      data: { data: [
        { id: '1', name: 'Product 1' }, 
        { id: '2', name: 'Product 2' },
        { id: '3', name: 'Product 3' }
      ] },
      refetch: mockRefetch,
    });

    (useMyContextProduct as jest.Mock).mockReturnValue({
      setProducts: mockSetProducts,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  })

  it('renders SearchInput component correctly', () => {
    const tree = renderer.create(<SearchInput placeholder="Search..." />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('displays the correct placeholder', () => {
    const instance = renderer.create(<SearchInput placeholder="Search..." />).root;
    const input = instance.findByType(TextInput);

    expect(input.props.placeholder).toBe('Search...');
  });

  it('updates searchText on text input change', () => {
    const instance = renderer.create(<SearchInput placeholder="Search..." />).root;
    const input = instance.findByType(TextInput);

    act(() => input.props.onChangeText('2'));

    expect(input.props.value).toBe('2');
  });
  
  it('sets returnKeyType to "done"', () => {
    const instance = renderer.create(<SearchInput placeholder="Search..." />).root;
    const input = instance.findByType(TextInput);

    expect(input.props.returnKeyType).toBe('done');
  });

});