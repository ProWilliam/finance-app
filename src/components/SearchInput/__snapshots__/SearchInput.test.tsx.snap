// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> renders SearchInput component correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "margin": 40,
      "paddingVertical": 10,
      "width": "80%",
    }
  }
>
  <TextInput
    keyboardType="number-pad"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search..."
    placeholderTextColor="black"
    returnKeyType="done"
    style={
      {
        "borderColor": "#eaeaea",
        "borderRadius": 8,
        "borderWidth": 1,
        "fontSize": 16,
        "height": 60,
        "overflow": "hidden",
        "paddingHorizontal": 20,
        "width": "100%",
      }
    }
    value=""
  />
</View>
`;
